compiler/runtime/OMRRSSReport.hpp

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+ * Copyright IBM Corp. and others 2000
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#ifndef OMR_RSSREPORT_INCL
+#define OMR_RSSREPORT_INCL
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#ifdef LINUX
+#include <unistd.h>
+#endif
+#include "env/TRMemory.hpp"
+#include "infra/Array.hpp"
+#include "infra/CriticalSection.hpp"
+#include "infra/List.hpp"
+#include "env/VerboseLog.hpp"
+#include "ras/DebugCounter.hpp"
+
+namespace TR { class Monitor; }
+
+namespace OMR
+{
+
+/*
+ * RSSItem describes content of some memory given address and size.
+ * RSSItem can be added to RSSRegion. It is possible to associate
+ * debug counters with each item.
+ */
+class RSSItem
+   {
+   public:
+   TR_ALLOC(TR_Memory::Inliner);  // TODO: add new type
+   enum Type
+      {
+      unknown,
+      header,
+      alignment,
+      coldBlocks,
+      overEstimate,
+      OOL,
+      snippet
+      };
+
+   static const char *itemNames[];
+
+   RSSItem() : _type(unknown), _addr(NULL), _size(0) {}
+   RSSItem(Type type, uint8_t *addr, size_t size, TR_PersistentList<TR::DebugCounterAggregation> counters,
+              const char* methodSignature = NULL) : _type(type), _addr(addr), _size(size), _counters(counters) {}
+
+   private:
+   enum Type _type;
+   uint8_t *_addr;
+   size_t _size;
+   TR_PersistentList<TR::DebugCounterAggregation> _counters;
```

Comment:
I actually thought that conceptually it works well:` RSSReport` is a global singleton (created in the persistent memory) that contains all the necessary data within itself or points to some persistent data. The TR_Persistent  data structures within it (list of regions, items and so on) hold the same assumption. I think the problem with the copy constructor for the TR_PersistentList within TR RSSItem will only arise if we start copying items. But we don't do it and if we do we actually might want to have a shallow copy (depends on why we would copy items).

Reply:
I think that in general everything is conceptually well defined; however, assuming that the `TR_PersistentList` won't accidentally be used or changed because `RSSReport` is a singleton is where I believe the conceptual integrity weakens. That coupling of assumptions shouldn't exist. All it would take, for example, is for someone to add a destructor that frees up memory to `TR_PersistentList` and this code would break.

Given that you said that `RSSReport` is a singleton, then perhaps the way around this is to have `RSSReport` have a list of `TR_PersistentList`s (where the first entry is a shared empty list). Then all you have to do, in say `OMR::X86::CodeGenerator::addItemsToRSSReport`, is something like

```
auto coldBlockCountersList = OMR::RSSReport::allocateNewPersistentList();
...
rssItem = new (self()->comp()->trPersistentMemory()) OMR::RSSItem(OMR::RSSItem::coldBlocks, coldCode, actualColdLength,
                                                                  coldBlockCountersList);
rssRegion->addRSSItem(rssItem, codeCache->getReservingCompThreadID(), methodName);
```
where `RSSReport::allocateNewPersistentList` may have a signature such as
```
TR_PersistentList<TR::DebugCounterAggregation>& RSSReport::allocateNewPersistentList();
```

This way the lifetimes of the persistent lists are explicitly tied to the `RSSReport` singleton, and there's no passing by value potential problems.

Reply:
I was actually thinking we I can improve the following:

1. **Encapsulation**: let OMR::RSSItem handle the storing of the debug counters and have a public method OMR::RSSItem::addDebugCounter(TR::DebugCounterAggregation *counterAggregation)
2. **Memory consumption:** not all items will have counters associated with them. Allocate counters on demand by having only a pointer in the item.

This will also make it very similar to other classes where we check if some list is null and allocate it if needed. However, generally, I would encourage using instance members more while providing all the necessary constructors and operators to the class. I think it's in the spirit of C++ to treat objects as values.

Reply:
Change instance variable to a pointer (for the counters) to save memory since some items will not have debug counters and we might have a lot of counters. Decided not to encapsulate the list since it makes it much more convenient to pass it as an argument when adding new item.